game-core/src/main/java/games/strategy/engine/data/UnitTypeList.java

```diff
@@ -40,6 +44,22 @@ public UnitType getUnitType(final String name) {
     return types;
   }
 
+  /**
+   * Returns the unit support rules for the unit types. Computed once and cached afterwards.
+   *
+   * @return The unit support rules.
+   */
+  public Set<UnitSupportAttachment> getSupportRules() {
```

Comment:
side-note, the method consuming this is overspecified
```
  private static void getSupport(
      final Collection<Unit> unitsGivingTheSupport,
      final Set<UnitSupportAttachment> rules,
      :
```

With the above, a consequence is if you have a `List` of rules, or any other non-Set collection, then you need to do a needless copy into a set. That method works fine when the collection type is changed to a Collection, ie:
```
  private static void getSupport(
      final Collection<Unit> unitsGivingTheSupport,
      final Collection<UnitSupportAttachment> rules,
     :
```

Reply:
I think it's a bit of a philosophical question - that applies to not just here. Would it be valid to repeat the same UnitSupportAttachment rule multiple times in the list? If logically that does not make sense, then to me it seems valid to use a Set to codify that in the API rather than a Collection which doesn't carry that guarantee when the function should not expect the same instance repeated in the collection.

Reply:
It's a matter of contract. A contract that accepts says "you must give me a set, otherwise there will be errors in how I do computation". Part of it is letting go and trusting the implementation that is passed is what the caller intends. If the caller is happy passing you duplicates, and you can process duplicates, then it is an issue with the implementation. If non-duplicates are needed for correctness, then it makes sense to require a set (at which point it's very justified to force callers to convert their list to a set). If duplicates are just inefficient, then it's on the caller for not doing the cleanup first, and you may be incurring worse efficiencies by forcing a collection copy. At the end of the day it's a trust that the implementation type satisfies any contract that you need. If you need unique items, then the contract should be set, if you need ordering, then list, if you really don't care (but set would be more efficient), then use collection and hope that callers give you a set (or handle it in the processing).

Reply:
From another perspective, it's simply the caller that often decides if duplicates are allowed or not. They instantiate the data, they likely use it in other places/ways - they probably are passing you a set. Otherwise, set should be required as an input when it is required for correctness, it enforces a contract. Needless enforcement between list and set often results in a lots of unnecessary data copy. When done well, using abstract types, often the concrete type is defined in just one place, can be changed, and everything just works with minimal to no copying of collection data.